### PR TITLE
island-ui/ ModalBase + DialogPrompt

### PR DESCRIPTION
**Pull request by DavidBachmann**
# Modal solutions

## What

Basic Modal foundation and DialogPrompt, a dialog that presents the user with a yes/no choice, that builds on BaseModal.

## Screenshots / Gifs

image

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against master before asking for a review

### CODE REVIEW

**Comment by DavidBachmann** [comment]
Discussion to be had before merging:

**Should we also provide a more basic modal component and build DialogPrompt on that?**
My opinion: Reakit's Dialog is a great building block for projects that need more advanced modals. island-ui shouldn't try to abstract that any more than Reakit is already doing.

File: libs/island-ui/core/src/lib/DialogPrompt/DialogPrompt.treat.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
+import { style } from 'treat'
+import { theme } from '@island.is/island-ui/theme'
+
+const hexToRgb = (hex: string) => {
```
Comment (jeremybarbet) [inline libs/island-ui/core/src/lib/DialogPrompt/DialogPrompt.treat.ts:4]: I've seen this in multiples places, I would move it to utils and export it from islandui

**Review by jeremybarbet — DISMISSED** [review]

File: libs/island-ui/core/src/lib/DialogPrompt/DialogPrompt.treat.ts (hunk 1 above)
Comment (DavidBachmann) [reply libs/island-ui/core/src/lib/DialogPrompt/DialogPrompt.treat.ts:4]: You're right. A util like
```
const hexToRgba = (hex: string, alpha: number) => {
  // Requires 6 number hexadecimal
  const rgb = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex)

  if (!rgb) {
    return null
  }

  const r = parseInt(rgb[1], 16)
  const g = parseInt(rgb[2], 16)
  const b = parseInt(rgb[3], 16)

  return `rgba(${r}, ${g}, ${b}, ${alpha})`
}
```

**Comment by barabrian** [comment]
> Discussion to be had before merging:
> 
> **Should we also provide a more basic modal component and build DialogPrompt on that?**
> My opinion: Reakit's Dialog is a great building block for projects that need more advanced modals. island-ui shouldn't try to abstract that any more than Reakit is already doing.

I agree Reakit is doing a-lot for us, but I think if there are more components going to use this pattern I think abstracting the core boilerplate and styles for a modal component could be useful, but that totally depends on what the designers have in mind 😄

File: libs/island-ui/core/src/lib/DialogPrompt/DialogPrompt.treat.ts (hunk 1 above)
Comment (jeremybarbet) [reply libs/island-ui/core/src/lib/DialogPrompt/DialogPrompt.treat.ts]: I would probably re-use where it was defined locally in each projects at the same time

Comment (DavidBachmann) [reply libs/island-ui/core/src/lib/DialogPrompt/DialogPrompt.treat.ts]: I don't want to touch other projects in this particular PR. Will make another for that.

**Comment by DavidBachmann** [comment]
> I agree Reakit is doing a-lot for us, but I think if there are more components going to use this pattern I think abstracting the core boilerplate and styles for a modal component could be useful, but that totally depends on what the designers have in mind 😄

Sold. I'll experiment with that.

File: libs/island-ui/core/src/lib/DialogPrompt/DialogPrompt.tsx [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+export const DialogPrompt = ({
+  title,
+  description,
+  ariaLabel,
+  baseId,
+  disclosureElement,
+  onConfirm,
+  onCancel,
+  buttonTextCancel = 'Cancel',
+  buttonTextConfirm = 'Confirm',
```
Comment (barabrian) [inline libs/island-ui/core/src/lib/DialogPrompt/DialogPrompt.tsx:44]: not sure about using defaults here, maybe we could use this to conditionally show the buttons, maybe some cases only have one button or no buttons

**Review by barabrian — COMMENTED** [review]
LGTM! I only have one comment we could discuss

File: libs/island-ui/core/src/lib/DialogPrompt/DialogPrompt.tsx [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+}: DialogPromptProps) => (
+  <ModalBase
+    disclosure={disclosureElement}
+    baseId={baseId}
+    aria-label={ariaLabel}
+    className={styles.dialog}
+  >
+    {({ closeModal }) => {
+      const handleClose = () => {
+        onCancel && onCancel()
```
Comment (jeremybarbet) [inline libs/island-ui/core/src/lib/DialogPrompt/DialogPrompt.tsx:54]: You can do 
```suggestion
        onCancel?.()
```

File: libs/island-ui/core/src/lib/DialogPrompt/DialogPrompt.tsx (hunk 2 above)
Comment (jeremybarbet) [reply libs/island-ui/core/src/lib/DialogPrompt/DialogPrompt.tsx:44]: I agree, we should be able to hide the buttons if needed. But I like having default values 😅

Comment (barabrian) [reply libs/island-ui/core/src/lib/DialogPrompt/DialogPrompt.tsx:44]: if we use defaults, wouldn't we need to make it translate-able?

Comment (jeremybarbet) [reply libs/island-ui/core/src/lib/DialogPrompt/DialogPrompt.tsx:44]: Could point, and we don't want to handle translation inside the island-ui, probably good to stick to what you said above then 👍

File: libs/island-ui/core/src/lib/DialogPrompt/DialogPrompt.tsx (hunk 3 above)
Comment (DavidBachmann) [reply libs/island-ui/core/src/lib/DialogPrompt/DialogPrompt.tsx:54]: Sorcery

File: libs/island-ui/core/src/lib/DialogPrompt/DialogPrompt.tsx (hunk 2 above)
Comment (DavidBachmann) [reply libs/island-ui/core/src/lib/DialogPrompt/DialogPrompt.tsx]: Addressed in https://github.com/island-is/island.is/pull/1474/commits/97880b29409df6fc36c69d4fdd0e8725a71c4a39

**Review by barabrian — APPROVED** [review]
Nice! LGTM

**Review by barabrian — APPROVED** [review]

**Pull request merged**